include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -992,6 +1012,63 @@ struct SessionOptions : detail::SessionOptionsImpl<OrtSessionOptions> {
+  using B = ConstModelCompilationOptionsImpl<T>;
+  using B::B;
+
+  ModelCompilationOptionsImpl& SetInputModelPath(const ORTCHAR_T* input_model_path);  ///< Wraps OrtApi::ModelCompilationOptions_SetInputModelPath
+  ModelCompilationOptionsImpl& SetInputModelFromBuffer(const void* input_model_data,
+                                                       size_t input_model_data_size);   ///< Wraps OrtApi::ModelCompilationOptions_SetInputModelFromBuffer
+  ModelCompilationOptionsImpl& SetEpContextEmbedMode(bool embed_ep_context_in_model);   ///< Wraps OrtApi::ModelCompilationOptions_SetEpContextEmbedMode
+  ModelCompilationOptionsImpl& SetOutputModelPath(const ORTCHAR_T* output_model_path);  ///< Wraps OrtApi::ModelCompilationOptions_SetOutputModelPath
+  ModelCompilationOptionsImpl& SetOutputModelExternalInitializersFile(const ORTCHAR_T* file_path,
+                                                                      size_t initializer_size_threshold);  ///< Wraps OrtApi::ModelCompilationOptions_SetOutputModelExternalInitializersFile
+  ModelCompilationOptionsImpl& SetOutputModelBuffer(OrtAllocator* allocator, void** output_model_buffer_ptr,
+                                                    size_t* output_model_buffer_size_ptr);  ///< Wraps OrtApi::ModelCompilationOptions_SetOutputModelBuffer
+};
+}  // namespace detail
+
+using UnownedModelCompilationOptions = detail::ModelCompilationOptionsImpl<detail::Unowned<OrtModelCompilationOptions>>;
+using ConstModelCompilationOptions = detail::ConstModelCompilationOptionsImpl<detail::Unowned<const OrtModelCompilationOptions>>;
+
+/** \brief Wrapper around ::OrtModelCompilationOptions
+ *
+ */
+struct ModelCompilationOptions : detail::ModelCompilationOptionsImpl<OrtModelCompilationOptions> {
+  explicit ModelCompilationOptions(std::nullptr_t) {}                              ///< Create an empty ModelCompilationOptions object, must be assigned a valid one to be used.
+  ModelCompilationOptions(const Env& env, const SessionOptions& session_options);  ///< Wraps OrtApi::CreateModelCompilationOptionsFromSessionOptions
+  ModelCompilationOptions(const Env& env, ConstSessionOptions session_options);    ///< Wraps OrtApi::CreateModelCompilationOptionsFromSessionOptions
+  explicit ModelCompilationOptions(OrtModelCompilationOptions* p) : ModelCompilationOptionsImpl<OrtModelCompilationOptions>{p} {}
+  UnownedModelCompilationOptions GetUnowned() const { return UnownedModelCompilationOptions{this->p_}; }
+  ConstModelCompilationOptions GetConst() const { return ConstModelCompilationOptions{this->p_}; }
+};
+
+/** \brief Compiles an input model to generate a model with EPContext nodes that execute EP-specific kernels. Wraps OrtApi::CompileModels.
+ *
+ * \param env: ORT environment object.
+ * \param model_compilation_options: Compilation options for a model.
+ * \return A Status indicating success or failure.
+ */
+Status CompileModel(const Env& env, ConstModelCompilationOptions model_compilation_options);
```

Comment:
why is the non-const non-reference variant needed?

Reply:
Out C++ API seems to be inconsistent with how we represent a constant handle to an opaque object. One way is using a `ConstMyClass` class and another is using a `const MyClass&`. We use both throughout the C++ API. Both mean the same thing, so want to provide overloads for both for convenience.

Reply:
Actually, I removed the overload with `ConstModelCompilationOptions` for now as that type is not really used.